File: leetcode/739-daily-temperatures/DailyTemperatures.java
import java.util.Stack;

/* * 
 * LC 739 - Daily Temperatures
 * 
 * Given an array of integers temperatures represents the daily temperatures, 
 * return an array answer such that answer[i] is the number of days you have to wait after the ith day to get a warmer temperature. 
 * 
 * If there is no future day for which this is possible, keep answer[i] == 0 instead.
*/
public class DailyTemperatures {
    public int[] dailyTemperatures(int[] temperatures) {
        int[] result = new int[temperatures.length];
        Stack<int[]> stack = new Stack<>(); // [temp, index]
        for (int i = 0; i < temperatures.length; i++) {
            int temp = temperatures[i];
            while (!stack.isEmpty() && temp > stack.peek()[0]) {
                int[] pair = stack.pop();
                result[pair[1]] = i - pair[1];
            }

            stack.push(new int[]{temp, i});
        }

        return result;
    }
}

File: leetcode/739-daily-temperatures/daily_temperatures.py
class Solution:
    """ 
    Given an array of integers temperatures represents the daily temperatures, 
    return an array answer such that answer[i] is the number of days you have to wait after the ith day to get a warmer temperature. 
    
    If there is no future day for which this is possible, keep answer[i] == 0 instead.
    """
    
    @classmethod
    def daily_temperatures(self, temperatures: list[int]) -> list[int]:
        result = [0] * len(temperatures)
        stack = []  # temp, index
        
        for i, t in enumerate(temperatures):
            while stack and t > stack[-1][1]:
                top_temp, top_index = stack.pop()
                result[top_index] = i - top_index
                
            stack.append((t, i))
        return result
    
def main():
    print(Solution.daily_temperatures(temperatures = [30, 40, 50, 60]))

if __name__ ==  "__main__":
    main()
    